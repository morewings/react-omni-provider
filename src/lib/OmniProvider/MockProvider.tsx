import type {FC, ReactNode} from 'react';

export const MockProvider: FC<{children: ReactNode; testId?: string; color?: string}> = ({children, testId, color}) => {
    return (
        <div data-testid={testId} style={{padding: '36px', backgroundColor: color}}>
            {children}
        </div>
    );
};
