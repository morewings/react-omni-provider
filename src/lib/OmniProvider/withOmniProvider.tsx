import type {ComponentType, FC, JSX} from 'react';

import {OmniProvider} from './OmniProvider.tsx';
import type {ProviderConfig} from './ProviderConfig.ts';

export const withOmniProvider =
    (providers: ProviderConfig[]) =>
    <TProps extends JSX.IntrinsicAttributes>(Component: ComponentType<TProps>) => {
        const result: FC<TProps> = props => (
            <OmniProvider providerConfig={providers}>
                <Component {...props} />
            </OmniProvider>
        );
        result.displayName = 'OmniProvider';
        return result;
    };
