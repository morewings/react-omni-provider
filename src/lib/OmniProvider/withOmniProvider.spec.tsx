import type {ComponentProps, FC} from 'react';
import {render} from '@testing-library/react';

import {MockProvider} from './MockProvider.tsx';
import {withOmniProvider} from './withOmniProvider.tsx';

const MockComponent: FC = () => {
    return <div data-testid="component" />;
};

describe('lib > withOmniProvider', () => {
    it('renders given component', () => {
        const Component = withOmniProvider([MockProvider])(MockComponent);
        const {getByTestId} = render(<Component />);
        expect(getByTestId('component')).toBeInTheDocument();
    });

    it('renders given provider', () => {
        const id = 'foo';
        const Component = withOmniProvider([[MockProvider, {testId: id} as ComponentProps<typeof MockProvider>]])(
            MockComponent
        );
        const {getByTestId} = render(<Component />);
        expect(getByTestId(id)).toBeInTheDocument();
    });

    it('respects left to right order of providers given', () => {
        const Component = withOmniProvider([
            [MockProvider, {testId: 'foo'} as ComponentProps<typeof MockProvider>],
            [MockProvider, {testId: 'bar'} as ComponentProps<typeof MockProvider>],
            [MockProvider, {testId: 'bazz'} as ComponentProps<typeof MockProvider>],
        ])(MockComponent);
        const {asFragment} = render(<Component />);
        expect(asFragment()).toMatchSnapshot();
    });
});
