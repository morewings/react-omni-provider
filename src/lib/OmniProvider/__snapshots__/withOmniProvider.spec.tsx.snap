// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib > withOmniProvider respects left to right order of providers given 1`] = `
<DocumentFragment>
  <div
    data-testid="foo"
    style="padding: 36px;"
  >
    <div
      data-testid="bar"
      style="padding: 36px;"
    >
      <div
        data-testid="bazz"
        style="padding: 36px;"
      >
        <div
          data-testid="component"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
